
Author: rmh
Status: gotta uncomment WARNS (see #351366), and find a portable way to add
  "-lbsd".  after this it can be sent upstream.

Index: usr.sbin/config/Makefile
===================================================================
--- usr.sbin/config/Makefile.orig
+++ usr.sbin/config/Makefile
@@ -9,11 +9,11 @@
 kernconf.c: kernconf.tmpl
 	file2c 'char kernconfstr[] = {' ',0};' < ${.CURDIR}/kernconf.tmpl > kernconf.c
 
-WARNS?=	6
-CFLAGS+= -I. -I${.CURDIR}
+#WARNS?=	6
+CFLAGS+= -I. -I${.CURDIR} -D_GNU_SOURCE
 
 DPADD=	${LIBL} ${LIBSBUF}
-LDADD=	-ll -lsbuf
+LDADD=	-ll -lbsd
 
 CLEANFILES+=	kernconf.c
 
Index: usr.sbin/config/main.c
===================================================================
--- usr.sbin/config/main.c.orig
+++ usr.sbin/config/main.c
@@ -41,12 +41,18 @@
   "$FreeBSD: src/usr.sbin/config/main.c,v 1.76 2007/05/17 04:53:52 imp Exp $";
 #endif /* not lint */
 
+#include <stdarg.h>
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <sys/sbuf.h>
 #include <sys/file.h>
 #include <sys/mman.h>
-#include <sys/param.h>
+#include <sys/param.h> /* BSD */
+
+#ifndef BSD
+#include <bsd/string.h> /* strlcpy, strlcat */
+#include "sbuf.c"
+#endif
 
 #include <assert.h>
 #include <ctype.h>
@@ -588,7 +594,11 @@
 	if ((dirp = opendir(p)) == NULL)
 		err(EX_OSERR, "opendir %s", p);
 	while ((dp = readdir(dirp)) != NULL) {
+#ifdef _DIRENT_HAVE_D_NAMLEN
 		i = dp->d_namlen - 2;
+#else
+		i = strlen (dp->d_name) - 2;
+#endif
 		/* Skip non-headers */
 		if (dp->d_name[i] != '.' || dp->d_name[i + 1] != 'h')
 			continue;
